File: api/src/main/java/org/apache/flink/agents/api/agents/Agent.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.agents.api.agents;

import org.apache.flink.agents.api.function.Function;
import org.apache.flink.agents.api.function.JavaFunction;
import org.apache.flink.agents.api.resource.ResourceDescriptor;
import org.apache.flink.agents.api.resource.ResourceType;
import org.apache.flink.agents.api.resource.SerializableResource;
import org.apache.flink.api.java.tuple.Tuple3;

import javax.annotation.Nullable;

import java.lang.reflect.Method;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.Map;

/** Base class for defining agent logic. */
public class Agent {
    private final Map<String, Tuple3<String[], Function, Map<String, Object>>> actions;

    private final Map<ResourceType, Map<String, Object>> resources;

    public Agent() {
        this.resources = new HashMap<>();
        for (ResourceType type : ResourceType.values()) {
            this.resources.put(type, new HashMap<>());
        }
        this.actions = new LinkedHashMap<>();
    }

    public Map<String, Tuple3<String[], Function, Map<String, Object>>> getActions() {
        return actions;
    }

    public Map<ResourceType, Map<String, Object>> getResources() {
        return resources;
    }

    /**
     * Add action to agent.
     *
     * @param triggerConditions Raw event-type names or Boolean conditions combined with OR
     *     semantics. Shape and expression validation occur during {@code AgentPlan} construction.
     * @param method The method of this action, should be static method.
     * @param config The optional config can be used by this action.
     */
    public Agent addAction(
            String[] triggerConditions, Method method, @Nullable Map<String, Object> config) {
        return addAction(
                method.getName(), triggerConditions, JavaFunction.fromMethod(method), config);
    }

    /**
     * Add action to agent.
     *
     * @param triggerConditions Raw event-type names or Boolean conditions combined with OR
     *     semantics. Shape and expression validation occur during {@code AgentPlan} construction.
     * @param method The method of this action, should be static method.
     */
    public Agent addAction(String[] triggerConditions, Method method) {
        return addAction(triggerConditions, method, null);
    }

    /**
     * Add action to agent.
     *
     * @param name The action name. Must be unique within this agent.
     * @param triggerConditions Raw event-type names or Boolean conditions combined with OR
     *     semantics. Shape and expression validation occur during {@code AgentPlan} construction.
     * @param function The api-layer function descriptor; will be promoted to a plan-layer
     *     executable at {@code AgentPlan} construction.
     * @param config Optional config for this action.
     */
    public Agent addAction(
            String name,
            String[] triggerConditions,
            Function function,
            @Nullable Map<String, Object> config) {
        if (actions.containsKey(name)) {
            throw new IllegalArgumentException(String.format("Action %s already defined.", name));
        }
        actions.put(name, new Tuple3<>(triggerConditions, function, config));
        return this;
    }

    public void addResourcesIfAbsent(Map<ResourceType, Map<String, Object>> resources) {
        for (ResourceType type : resources.keySet()) {
            Map<String, Object> typedResources = resources.get(type);
            typedResources.forEach(this.resources.get(type)::putIfAbsent);
        }
    }

    /**
     * Add resource to agent.
     *
     * @param name The name indicate the resource.
     * @param type The type of the resource.
     * @param instance The serializable resource object, the resource descriptor, or — for an {@code
     *     AGENT} resource — an {@link Agent} to compile into an internal sub-agent.
     */
    public Agent addResource(String name, ResourceType type, Object instance) {
        if (resources.get(type).containsKey(name)) {
            throw new IllegalArgumentException(String.format("%s %s already defined.", type, name));
        }
        checkNoChatModelRouterNameClash(name, type, resources);

        if (instance instanceof SerializableResource) {
            resources.get(type).put(name, instance);
        } else if (instance instanceof ResourceDescriptor) {
            resources.get(type).put(name, instance);
        } else if (instance instanceof Agent) {
            resources.get(type).put(name, instance);
        } else {
            throw new IllegalArgumentException(
                    String.format("Unsupported resource %s", instance.getClass().getName()));
        }
        return this;
    }

    /**
     * Chat models and model routers share the chat request namespace ({@code ChatRequestEvent}
     * names either), so one name must not be registered as both. Checked here, at the registration
     * call site, so the failure points at the user's own {@code addResource} line; {@code
     * AgentPlan} re-validates as a backstop.
     */
    public static void checkNoChatModelRouterNameClash(
            String name, ResourceType type, Map<ResourceType, Map<String, Object>> resources) {
        ResourceType clashing =
                type == ResourceType.CHAT_MODEL
                        ? ResourceType.MODEL_ROUTER
                        : type == ResourceType.MODEL_ROUTER ? ResourceType.CHAT_MODEL : null;
        if (clashing != null
                && resources.containsKey(clashing)
                && resources.get(clashing).containsKey(name)) {
            throw new IllegalArgumentException(
                    String.format(
                            "'%s' is already registered as %s; chat models and model routers share the"
                                    + " chat request namespace and must use distinct names.",
                            name, clashing));
        }
    }

    public enum ErrorHandlingStrategy {
        FAIL("fail"),
        RETRY("retry"),
        IGNORE("ignore");

        private final String value;

        ErrorHandlingStrategy(String value) {
            this.value = value;
        }

        public String getValue() {
            return value;
        }
    }

    public static String STRUCTURED_OUTPUT = "structured_output";
}

File: python/flink_agents/api/agents/agent.py
################################################################################
#  Licensed to the Apache Software Foundation (ASF) under one
#  or more contributor license agreements.  See the NOTICE file
#  distributed with this work for additional information
#  regarding copyright ownership.  The ASF licenses this file
#  to you under the Apache License, Version 2.0 (the
#  "License"); you may not use this file except in compliance
#  with the License.  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
# limitations under the License.
#################################################################################
from abc import ABC
from typing import Any, Callable, Dict, List, Tuple, Union

from flink_agents.api.function import Function, PythonFunction
from flink_agents.api.resource import (
    ResourceDescriptor,
    ResourceType,
    SerializableResource,
)

STRUCTURED_OUTPUT = "structured_output"


class Agent(ABC):
    """Base class for defining agent logic.


    Example:
        Users have two ways to create an Agent

        * Declare an Agent with decorators
        ::

            class MyAgent(Agent):
                @action(EventType.InputEvent)
                @staticmethod
                def my_action(event: Event, ctx: RunnerContext) -> None:
                    action logic

                @chat_model_connection
                @staticmethod
                def my_connection() -> ResourceDescriptor:
                    return ResourceDescriptor(clazz=OllamaChatModelConnection,
                                              model="qwen2:7b",
                                              base_url="http://localhost:11434")

                @chat_model_setup
                @staticmethod
                def my_chat_model() -> ResourceDescriptor:
                    return ResourceDescriptor(clazz=OllamaChatModel,
                                              connection="my_connection")

        * Add actions and resources to an Agent instance
        ::

            my_agent = Agent()
            my_agent.add_action(name="my_action",
                                trigger_conditions=["_input_event"],
                                func=action_function)
                    .add_resource(name="my_connection",
                                  instance=ResourceDescriptor(
                                        clazz=OllamaChatModelConnection,
                                        arg1=xxx
                                )
                    .add_resource(
                        name="my_connection",
                        instance=ResourceDescriptor(
                            clazz=OllamaChatModelConnection,
                            arg1=xxx
                        )
                    )
                    .add_resource(
                        name="my_chat_model",
                        instance=ResourceDescriptor(
                            clazz=OllamaChatModelSetup,
                            connection="my_connection"
                        )
                    )
    """

    _actions: Dict[str, Tuple[List[str], Function, Dict[str, Any] | None]]
    _resources: Dict[ResourceType, Dict[str, Any]]

    def __init__(self) -> None:
        """Init method."""
        self._actions = {}
        self._resources = {}
        for type in ResourceType:
            self._resources[type] = {}

    @property
    def actions(
        self,
    ) -> Dict[str, Tuple[List[str], Function, Dict[str, Any] | None]]:
        """Get added actions."""
        return self._actions

    @property
    def resources(self) -> Dict[ResourceType, Dict[str, Any]]:
        """Get added resources."""
        return self._resources

    def add_action(
        self,
        name: str,
        trigger_conditions: List[str],
        func: Callable | Function,
        **config: Any,
    ) -> "Agent":
        """Add action to agent.

        Parameters
        ----------
        name : str
            The name of the action, should be unique in the same Agent.
        trigger_conditions : list[str]
            Raw event-type names or Boolean condition expressions combined with
            OR semantics. Expression validation occurs when the plan is applied.
        func : Callable | Function
            Either a raw Python callable (it will be wrapped as a
            ``PythonFunction``) or a pre-built flink-agents ``Function``
            (e.g. from the YAML loader).
        **config : Any
            Key named arguments can be used by this action in runtime.

        Returns:
        -------
        Agent
            The modified Agent instance.
        """
        if name in self._actions:
            msg = f"Action {name} already defined"
            raise ValueError(msg)
        if not isinstance(func, Function):
            func = PythonFunction.from_callable(func)
        self._actions[name] = (trigger_conditions, func, config if config else None)
        return self

    def add_resource(
        self,
        name: str,
        resource_type: ResourceType,
        instance: Union[SerializableResource, ResourceDescriptor, "Agent"],
    ) -> "Agent":
        """Add resource to agent instance.

        Parameters
        ----------
        name : str
            The name of the prompt, should be unique in the same Agent.
        resource_type: ResourceType
            The type of the resource.
        instance: SerializableResource | ResourceDescriptor | Agent
            The serializable resource instance, the descriptor of resource,
            or an Agent instance. For the AGENT resource type an ``Agent`` is
            compiled into an internal sub-agent during plan construction.

        Returns:
        -------
        Agent
            The agent to add the resource.
        """
        if name in self._resources[resource_type]:
            msg = f"{resource_type.value} {name} already defined"
            raise ValueError(msg)

        self._resources[resource_type][name] = instance
        return self
